### PR TITLE
Fixed shadowing and function names having a type.

### LINKED ISSUE #144: The type of a function's name is its return type, not function
Currently this will compile:
```
state Alive (0, 0, 0) {
    let x = f + f;
}

function f () int {
    return 0;
}
```
Because the type of "f" is "int". This is caused by this line: https://github.com/d409f19/cellmata/blob/master/compiler/src/main/kotlin/ast/symboltable.kt#L95

### PR DESCRIPTION
**Pull request by NicEastvillage**
See issue #144 for details. While fixing this I discovered that shadowing didn't work when it's type changed.
Resolves #144

### CODE REVIEW

File: compiler/src/main/kotlin/visitors/typechecker.kt [hunk 1]
```diff
@@ -324,7 +325,22 @@ class TypeChecker(symbolTable: Table) : ScopedASTVisitor(symbolTable = symbolTab
...
+            ErrorLogger += TypeError(node.ctx, "Found function name '${node.spelling}', but the function is not called.")
+        } else {
+            node.setType(
+                when (decl) {
+                    is StateDecl -> StateType
+                    is NeighbourhoodDecl -> LocalNeighbourhoodType
+                    is AssignStmt -> decl.expr.getType()
+                    is ConstDecl -> decl.expr.getType()
+                    is FunctionArgument -> decl.type
+                    else -> throw AssertionError()
```
Comment (falkecarlsen) [inline compiler/src/main/kotlin/visitors/typechecker.kt]: Why no error message for this?

**Review by falkecarlsen — CHANGES_REQUESTED** [review]
Looks good, just one question.

File: compiler/src/main/kotlin/visitors/typechecker.kt (hunk 1 above)
Comment (NicEastvillage) [reply compiler/src/main/kotlin/visitors/typechecker.kt]: It should never happen. Guess I can give it an error message anyway, so if we change something, we get a nice warning.

Comment (NicEastvillage) [reply compiler/src/main/kotlin/visitors/typechecker.kt]: It should never happen. Guess I can give it an error message anyway, so if we change something, we get a nice warning.

**Review by falkecarlsen — APPROVED** [review]

**Pull request merged**